### PR TITLE
Split opencl glm function

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

Splits OpenCL implementation of `normal_id_glm_lpdf()` into a separate function. This fixes side effect of PR #1353.

## Tests
New tests were added as CPU tests can't be reused on a function with different signature.

## Side Effects
None.

## Checklist

- [x] Math issue #1319

- [x] Copyright holder: Tadej Ciglarič (University of Ljubljana)

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
So in short you took the currently ifdefed implementation inside /prim to a separate file and presume that x and y are `matrix_cl`, which means that they cant be parameters. Anything else of note?

**Comment by t4c1** [comment]
No, that is all.

File: test/unit/math/rev/mat/prob/normal_id_glm_lpdf_gpu_test.cpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
+#ifdef STAN_OPENCL
```
Comment (rok-cesnovar) [inline test/unit/math/rev/mat/prob/normal_id_glm_lpdf_gpu_test.cpp]: Please move the file under `/opencl/rev/` You can omit the `_gpu_` then. Also, in general we use the `_cl` suffix for everything OpenCL related.

File: stan/math/opencl/normal_id_glm_lpdf.hpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/multiply.hpp>
+
+#include <cmath>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns the log PDF of the Generalized Linear Model (GLM)
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply stan/math/opencl/normal_id_glm_lpdf.hpp]: Also add the include to stan/math/opencl/normal_id_glm_lpdf.hpp to opencl.hpp

File: stan/math/opencl/normal_id_glm_lpdf.hpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
+#ifndef STAN_MATH_OPENCL_NORMAL_ID_GLM_LPDF_HPP
+#define STAN_MATH_OPENCL_NORMAL_ID_GLM_LPDF_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/scal/err/check_consistent_sizes.hpp>
+#include <stan/math/prim/scal/err/check_finite.hpp>
```
Comment (rok-cesnovar) [inline stan/math/opencl/normal_id_glm_lpdf.hpp]: You are missing the include for `check_positive_finite`

File: stan/math/opencl/normal_id_glm_lpdf.hpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+ * @param sigma (Sequence of) scale parameters for the normal
+ * distribution.
+ * @return log probability or log sum of probabilities
+ * @throw std::domain_error if x, beta or alpha is infinite.
+ * @throw std::domain_error if the scale is not positive.
+ * @throw std::invalid_argument if container sizes mismatch.
+ */
+template <bool propto, typename T_alpha, typename T_beta, typename T_scale>
+return_type_t<T_alpha, T_beta, T_scale> normal_id_glm_lpdf(
+    const matrix_cl<double> &y_cl, const matrix_cl<double> &x_cl,
```
Comment (seantalts) [inline stan/math/opencl/normal_id_glm_lpdf.hpp]: This looks good to me from the stanc3 perspective! We'll need to specify in the type signatures list which arguments are required to be opencl if the fn is to execute on opencl so we know which ones to promote.

File: test/unit/math/rev/mat/prob/normal_id_glm_lpdf_gpu_test.cpp [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+  matrix_cl<double> x_cl(x);
+  matrix_cl<double> x_size1_cl(x_size1);
+  matrix_cl<double> x_size2_cl(x_size2);
+  matrix_cl<double> x_value_cl(x_value);
+  matrix_cl<double> y_cl(y);
+  matrix_cl<double> y_size_cl(y_size);
+  matrix_cl<double> y_value_cl(y_value);
+
+  EXPECT_NO_THROW(
+      stan::math::normal_id_glm_lpdf(y_cl, x_cl, alpha, beta, sigma));
```
Comment (rok-cesnovar) [inline test/unit/math/rev/mat/prob/normal_id_glm_lpdf_gpu_test.cpp]: Also add tests that check the behavior for size zero. I dont see an equivalent of 
```
if (size_zero(y, x, beta, sigma)) {	  if (size_zero(y, x, beta, sigma)) {
    return 0;	    return 0;
  }
```
in the gpu glm, so I am guessing it doesnt work for that case? I did go and check that none of the GLMs have size zero tests which is not great. I will add an issue.

Comment (t4c1) [reply test/unit/math/rev/mat/prob/normal_id_glm_lpdf_gpu_test.cpp]: In most functions this is just an optimization for special case. However an opencl kernel can not be launched with global size of 0, so I'm adding such check and tests.

**Comment by t4c1** [comment]
Yay, auto formatting is back!

**Pull request merged**